### ISSUE TITLE
Syntax changes to type signatures

### ISSUE THREAD

**Issue by dorchard** [dorchard/granule#10]
@vimuel  and I discussed moving from syntax like `|type| coeffect` to `type |coeffect|`, e.g., 
instead of:
```
dub : |Int| 2 -> Int
dub |x| = x + x

twice : forall (c : Nat) . ||Int| c -> Int| 2 -> |Int| (2 * c) -> Int
twice |g| |x| = g |x| + g |x|
```
write this instead:
```
dub : Int |2| -> Int
dub |x| = x + x

twice : forall (c : Nat) . (Int |c| -> Int) |2| -> Int |2 * c| -> Int
twice |g| |x| = g |x| + g |x|
```
Formally, this is less ambiguous, but I think is humanly more clear too (e.g. you don't have to think about whether you want a coeffect type till after you've written the core type).

**Comment by dorchard** [comment]
Parser change would be:
```
     | Type '|' Coeffect '|'       { Box $3 $1 }
```
(need to change all the examples).

**Comment by heades** [comment]
I agree, and think this is a better syntax.

**Comment by vimuel** [comment]
Done. Now what about effects?

Would it make sense to change this:
```
doTwice : (Int -> <Int> [R]) |2| -> Int |2| -> <Int> [R, R]
doTwice |f| |x| = let <a : Int> = f x in
                  let <b : Int> = f x in
                  <a + b>
```
to this
```
doTwice : (Int -> Int <R>) |2| -> Int |2| -> Int <R, R>
doTwice |f| |x| = let <a : Int> = f x in
                  let <b : Int> = f x in
                  <a + b>
```
?

**Comment by dorchard** [comment]
I guess technically it should be:
```
doTwice : (Int -> Int <[R]>) |2| -> Int |2| -> Int <[R, R]>
doTwice |f| |x| = let <a : Int> = f x in
                  let <b : Int> = f x in
                  <a + b>
```
As the grade is a list of effect tokens.

**Comment by dorchard** [comment]
We probably also need to change the Curry-style typing in `let`.
Currently:
`let |x : Int| Nat = xBox in e`
but perhaps it should be more like:
`let |x| : Int |Nat| = xBox in e`
Or perhaps we need to fight to do away with these signatures in a `let` anyway, which should be possible with a reformulation of the bidirectional rules.

**Comment by vimuel** [comment]
Adding `let |x| : Int |Nat| = xBox in e` seems to lead to an ambiguous grammar due to `| Type '|' Coeffect '|'`. Let's discuss tomorrow morning.

**Comment by dorchard** [comment]
This was fixed by 366752f3b9d728d91e21fb5d7f7c15dc70537c82.
